DEVELOPER_GUIDE.md

```diff
@@ -92,19 +91,21 @@ sourceCompatibility = JavaVersion.VERSION_11
 targetCompatibility = JavaVersion.VERSION_11
 ```
 
-Download Java 11 from [here](https://adoptium.net/releases.html?variant=openjdk11).
+Download Java 11 from [here](https://adoptium.net/releases.html?variant=openjdk11). 
 
-#### JDK 14
+Since 8.10 release, Gradle has deprecated the usage of the JDKs below JDK-16. As such, for development purposes we recommend to install additionally at least [JDK 17](https://adoptium.net/temurin/releases/?version=17) or [JDK 21](https://adoptium.net/temurin/releases/?version=21). If you still want to stick to JDK-11 only, please add `-Dorg.gradle.warning.mode=none` when invoking any Gradle build task from command line, for example:
 
-To run the full suite of tests, download and install [JDK 14](https://jdk.java.net/archive/) and set `JAVA11_HOME`, and `JAVA14_HOME`. They are required by the [backwards compatibility test](./TESTING.md#testing-backwards-compatibility).
+```
+./gradlew check -Dorg.gradle.warning.mode=none
+```
 
-#### JDK 17
```

Comment:
Should we add JDK17 back as @andrross called out we need 17 for bwc?

Reply:
On the second thought, I think JDK-17 should probably stay there, thanks @owaiskazi19 !

Reply:
I'm wondering if we could simplify the three "JDK" headings down to just this:

> \#### JDK
> 
> OpenSearch requires a minimum of JDK-11 in order to build. However, it is recommended to use a more recent LTS version of the JDK and set it to the `JAVA_HOME` environment variable. In addition, certain backward compatibility tests check out and compile the previous major version of OpenSearch, and therefore require installing JDK-11 and JDK-17 and setting the `JAVA11_HOME` and `JAVA17_HOME` environment variables.
> 
> By default, the test tasks use bundled JDK at runtime, which is configured in version catalog [gradle/libs.versions.toml](gradle/libs.versions.toml).

The existing text has a lot of detail that I think is maybe not necessary?

Reply:
> I'm wondering if we could simplify the three "JDK" headings down to just this:
> 
> > #### JDK
> > OpenSearch requires a minimum of JDK-11 in order to build. However, it is recommended to use a more recent LTS version of the JDK and set it to the `JAVA_HOME` environment variable. In addition, certain backward compatibility tests check out and compile the previous major version of OpenSearch, and therefore require installing JDK-11 and JDK-17 and setting the `JAVA11_HOME` and `JAVA17_HOME` environment variables.
> > By default, the test tasks use bundled JDK at runtime, which is configured in version catalog [gradle/libs.versions.toml](gradle/libs.versions.toml).
> 
> The existing text has a lot of detail that I think is maybe not necessary?

There are a few things here that do not play well (hence this pull request):
 - due to Gradle, to build with any JDK below 16, one needs to add `-Dorg.gradle.warning.mode=none`
 - due to Gradle, build with any JDK above 23 will fail  (like 24-ea for example)

So I think we should be very explicit what we support or not, so I could introduce the build matrix instead. Also, the large chunk of this work will be throwaway very soon (after https://github.com/opensearch-project/OpenSearch/pull/16366). 

If you don't mind, I would prefer to keep this guide exact (with respect to JDK versions) and get back to it once we level up to JDK-21.

Reply:
Sure, that's fair, but what I don't like about the text in this PR is towards the beginning of this section it says:

> This means you must have a JDK 11 installed with the environment variable `JAVA_HOME` referencing the path to Java home for your JDK 11 installation, e.g. `JAVA_HOME=/usr/lib/jvm/jdk-11`.

Only if you keep reading does it recommend to use Java 17 or 21. And nowhere does it mention the JAVA11_HOME and JAVA17_HOME environment variables which are required for bwc tests. I also don't love the snippets of gradle configuration which is destined to be out of date when things change anyway.

Reply:
Valid points, reworked these sections, thanks @andrross !